### DISCUSSION TITLE
Instanced pipelines break build history

### DISCUSSION THREAD

**Discussion by mvdkleijn** [concourse/concourse#6869] [Help & Support]
## Summary

The build history of instanced pipelines is lost when variable values change due to the fact that those are considered "new" pipelines.

## Steps to reproduce

Create an instanced pipeline with var=A
Run pipeline
Change var=B
Witness previous build's history is gone

## Expected results

Instanced pipeline can be assigned an ID so the build history is retained.

## Actual results

Build history bye bye.

## Additional context

-

## Triaging info

* Concourse version: 7.1.0
* Browser (if applicable): -
* Did this used to work?

**Comment by taylorsilva** [comment]
Hey, so the pipelines are now identified by their "group" name and all their vars. This is how it was described in the RFC https://github.com/concourse/rfcs/blob/master/034-instanced-pipelines/proposal.md
and explained in the docs https://concourse-ci.org/instanced-pipelines.html

I'm guessing you got hit by the automatic archiving and were using the `set_pipeline` step to set you instance pipelines?
https://github.com/concourse/rfcs/blob/master/033-archiving-pipelines/proposal.md

I think a workaround for you right now would be to add to the list of pipelines you're setting with `set_pipeline` instead of updating the vars for the pipeline. Then the "old" pipeline won't be archived and it's build history will be viewable still.

**Comment by mvdkleijn** [comment]
Hey there @taylorsilva thanks for the reply. I actually hadn't realized that there is automatic archiving going on. This might indeed explain it, I'll have to check that tomorrow.

What we're doing is building Docker containers with specific versions of apps inside them. The version is used as a variable in the pipeline instance. This would mean there are only a few builds in each instance and potentially lots of pipelines in the scenario that you describe. Not an ideal case.

I quite liked the idea of having instanced pipelines with the version variable shown in the UI... it would allow us to see the versions of the various apps in once glance. However, if this is impossible, I guess I could try a sub-optimal route and dig the version out of a file or hardcode it in the Dockerfile or something.

**Comment by taylorsilva** [comment]
Yes, I see how that would get noisy really fast. I would probably try and add whatever app(s) version you're tracking as resources to your pipeline instead. That feels like the more concourse way to interact with the app versions as inputs to your pipelines. Then build history would persist.

**Comment by mvdkleijn** [comment]
@taylorsilva Hi there, I just realized something... we're using instanced pipelines and are trying to set webhooks on them... However, the fact that they don't have a consistent name is plaguing us. Is there a way to trigger a resource check in an instanced pipeline that does not require us to use the variable fields in the pipeline name?

**Reply by taylorsilva** [reply]
Going to ping @aoldershaw to see if he has any suggestions 😬

**Reply by aoldershaw** [reply]
Don't have any suggestions really. To be honest, this isn't a use case we really considered when designing instanced pipelines.

The intention behind instanced pipelines was that the instance vars are part of the identity of the pipeline, so if you have different instance vars, then it's a completely different pipeline. Instanced pipelines with the same name will be grouped together visually in the UI to make it easier to organize them, but they are just distinct pipelines.

Trying to understand your use case a little better - is there always just 1 instance at a given time for the latest versions of the bundled apps? If so (and correct me if I'm wrong), I think the better way to model this would just be to have a single pipeline where the bundled app versions are modelled as resources. Is it fair to say you're mainly using instanced pipelines as a way to get the UI to display the latest versions of these resources (rather than needing to click on each of the resources individually)?

If so, maybe we could find a new UI design to make it easier to see the versions of resources at a glance. Or if the app versions can't easily be modelled as resources, maybe there's a way to tag pipelines with metadata that we could display somewhere in the UI.

What do you think of these suggestions?  Let me know if I'm off base with my interpretation!

**Reply by mvdkleijn** [reply]
@aoldershaw Thanks for the reply, I'll attempt to give a clear explanation. 😄 

> Don't have any suggestions really. To be honest, this isn't a use case we really considered when designing instanced pipelines.
> 
> The intention behind instanced pipelines was that the instance vars are part of the identity of the pipeline, so if you have different instance vars, then it's a completely different pipeline. Instanced pipelines with the same name will be grouped together visually in the UI to make it easier to organize them, but they are just distinct pipelines.

I understand the use case and the intention. I'm not sure it matches with what we need though. There are two main issues that I see... one is that I'd like to be able to name instanced pipelines, which is admittedly probably a nice to have.

And two, more serious in my opinion, that it appears to be very difficult if not impossible to use webtokens.
 
> Trying to understand your use case a little better - is there always just 1 instance at a given time for the latest versions of the bundled apps? If so (and correct me if I'm wrong), I think the better way to model this would just be to have a single pipeline where the bundled app versions are modelled as resources. Is it fair to say you're mainly using instanced pipelines as a way to get the UI to display the latest versions of these resources (rather than needing to click on each of the resources individually)?

Concretely, the webtoken use case in our situation is as follows:

1. For various reasons, we need to use notification / webhooks instead of relying on the standard Concourse polling mechanism
2. We have a single template for (instanced) pipelines that we use to build various Docker containers with. The instance variables might be things like: git repo, base image name, base image version, java version, app version, etc.
3. If we want to trigger pipeline because a commit was made for example, we'd need to have the external system make a call to 

/api/v1/teams/TEAM_NAME/pipelines/PIPELINE_NAME/resources/RESOURCE_NAME/check/webhook?webhook_token=WEBHOOK_TOKEN

The webhook url contains the pipeline_name, so including the variables. This means that if any of the variables change, like a base image version, I'd have to update the external system's notification system to update the URL. Not even mentioning the fact that the pipeline_name is very unwieldy.

This effectively means that (for our use case at least) webhooks are broken when using instanced pipelines.

Why can't we use resources for app versions for example?

1. Because they are different apps, maintained by various external parties, some available as a downloadable http resource only which means we cannot generate available version
2. We want to have predictable versions in our containers instead of "latest", whatever that may be.

> If so, maybe we could find a new UI design to make it easier to see the versions of resources at a glance. Or if the app versions can't easily be modelled as resources, maybe there's a way to tag pipelines with metadata that we could display somewhere in the UI.

Some of the things might be UI related, but (in my opinion) this is mostly a matter of us wanting to reuse the same pipeline (template) with a couple of minor differences... the whole point of instanced pipelines, right?

I hope I'm clear with my explanations! 😄 

> What do you think of these suggestions? Let me know if I'm off base with my interpretation!

See above 😄

**Reply by aoldershaw** [reply]
I think webhooks generally are a pain in Concourse, but definitely more so with instanced pipelines (where the identity of the pipeline is dynamic by design). There's an [incomplete proposal](https://github.com/vito/rfcs/blob/webhooks/000-webhooks/proposal.md) for cluster-wide webhook support, but it's still a ways off.

> Some of the things might be UI related, but (in my opinion) this is mostly a matter of us wanting to reuse the same pipeline (template) with a couple of minor differences... the whole point of instanced pipelines, right?

To me, instanced pipelines are more about grouping related pipelines together that (typically) share a template. If you don't care about grouping (because there's only a single pipeline at a time), and you want to keep history between app version combinations, then perhaps instanced pipelines aren't the solution - there's nothing stopping you from using a pipeline template to continually set a "regular" pipeline.

One thing I'm still not sure about is whether there's always just one instanced pipeline that you intend to keep rolling forward? Also, how is the pipeline being set? i.e. are you setting it manually with `fly set-pipeline ... -i app1=1.2.3 -i app2=2.3.4`, or is this being automated somehow?

w.r.t displaying better information in the UI, we could possibly add a `display.metadata` field in the pipeline config:
```yaml
display:
  metadata:
    app1: ((app1_version))
    app2: ((app2_version))
...
```

...which could be displayed somewhere in the UI

**Reply by aoldershaw** [reply]
@mvdkleijn re: the difficulty you're running into with webhooks - I put together an RFC that allows for decoupling webhooks from resources, so you can e.g. configure a single webhook globally in your GitHub org, and have Concourse delegate to the correct resource(s) when webhook payloads come in. Would love any feedback: https://github.com/concourse/rfcs/pull/110
